src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/DocumentPullDiagnosticsEndpoint.cs

```diff
@@ -160,4 +165,39 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalDocumentDiagno
 
         return (delegatedResponse.CSharpDiagnostics, delegatedResponse.HtmlDiagnostics);
     }
+
+    /// <summary>
+    /// Reports telemetry for RZ10012 "Found markup element with unexpected name" to help track down potential issues
+    /// with taghelpers being discovered (or lack thereof)
+    /// </summary>
+    private async ValueTask ReportRZ10012TelemetryAsync(DocumentContext documentContext, VSInternalDiagnosticReport[]? razorDiagnostics, CancellationToken cancellationToken)
+    {
+        if (razorDiagnostics is null)
+        {
+            return;
+        }
+
+        if (_telemetryReporter is null)
+        {
+            return;
+        }
+
+        var relaventDiagnostics = razorDiagnostics.SelectMany(r => r.Diagnostics?.Where(d => d.Code == "RZ10012") ?? []).ToImmutableArray();
```

Comment:
💡 Rather than allocating a new array just to retrieve the length, consider using a combination of LINQ's `Sum()` and `Count(...)` and operators.

```suggestion
        var relevantDiagnosticCount = razorDiagnostics.Sum(static r => r.Diagnostics.Count(static d => d.Code == "RZ10012"));
        var relaventDiagnostics = razorDiagnostics.SelectMany(r => r.Diagnostics?.Where(d => d.Code == "RZ10012") ?? []).ToImmutableArray();
```